proposals/0035-linalg-matrix.md

```diff
@@ -47,70 +47,816 @@ concept ArithmeticScalar = std::is_arithmetic<T>::value;
+    uint Align = sizeof(__detail::ComponentTypeTraits<ComponentTy>::Type));
+
+static Matrix Matrix::Load(
+    RWByteAddressBuffer Res, uint StartOffset, uint Stride, bool ColMajor,
+    uint Align = sizeof(__detail::ComponentTypeTraits<ComponentTy>::Type));
+
+template <typename T>
+  requires ArithmeticScalar<T>
+static Matrix Matrix::Load(groupshared T Arr[], uint StartIdx, uint Stride,
+                           bool ColMajor);
+```
+
+The matrix `Load` methods create a new matrix of the specified dimensions and
+fill the matrix by reading data from the supplied source. Matrices can be read
+from `[RW]ByteAddressBuffer` objects or `groupshared` arrays. When read from
+`[RW]ByteAddressBuffer` objects the data is assumed to already be in the
+expected target data format. When read from `groupshared` memory, the data may
+be in any arithmetic or packed data type. If the type mismatches the target data
+type of the matrix a data conversion is applied on load.
+
+
+#### Matrix::Store
+
+```c++
+void Matrix::Store(
+    RWByteAddressBuffer Res, uint StartOffset, uint Stride, bool ColMajor,
+    uint Align = sizeof(__detail::ComponentTypeTraits<ComponentTy>::Type));
+
+template <typename T>
+  requires ArithmeticScalar<T>
+void Matrix::Store(groupshared T Arr[], uint StartIdx, uint Stride,
+                   bool ColMajor);
+```
+
+The matrix `Store` methods store the matrix data to a target
+`RWByteAddressBuffer` or `groupshared` array. When storing to
+`RWByteAddressBuffer` objects the data is stored in the component type of the
+matrix object. When storing to `groupshared` memory, the matrix component data
+is converted to the target arithmetic or packed data type if the data types do
+not match.
+
+#### Matrix::GetRow(uint)
+
+```c++
+vector<ElementType, M> Matrix::GetRow(uint Index);
+```
+
+Returns a row vector of the matrix as a vector of the underlying HLSL native
+element type. If `Index` is out of range for the matrix size the result is a `0`
+filled vector.
+
+
+#### Matrix::SetRow(vector<ElementType, M>, uint)
+
+```c++
+void Matrix::SetRow(vector<ElementType, M> V, uint Index);
+```
+
+Sets the specified matrix row to the value in the vector V. If the matrix scope
+is `Wave`, this behaves as if called as `SetRow(WaveReadLaneFirst(V), Index)`.
+If the `Index` is out of range of the matrix, this is a no-op.
+
+#### Matrix::Get(uint2)
+
+```c++
+std::enable_if_t<__detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar,
+                 ElementType>
+Matrix::Get(uint2 Index);
+```
+
+Accesses a specific component of the matrix using two-dimensional indexing. This
+method is only available if the component type has has native scalar support in
+HLSL. If the `Index` parameter is out-of range for the matrix the result is `0`
+casted to `ElementType`.
+
+#### Matrix::Set(ElementType, uint2)
+
+```c++
+std::enable_if_t<__detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar,

```

Comment:
Setting and getting are only available when the component has a native scalar type.
I imagine the use case for getting and setting is still interesting.

In future should there be getter and setter functionality in the additional cases, with implied conversions to a scalar HLSL type?